inc/namespace.php

```diff
@@ -0,0 +1,335 @@
+ * @param string $handle Unique asset handle.
+ * @return string|null Integrity hash if set, null otherwise.
+ */
+function get_hash_for_asset( WP_Dependencies $dependencies, string $handle ) : ?string {
+	return $dependencies->get_data( $handle, INTEGRITY_DATA_KEY ) ?? null;
+}
+
+/**
+ * Set the integrity hash for a script.
+ *
+ * @param string $handle Unique script handle.
+ * @param string $hash Integrity hash (in format "<algo>-<hash>").
+ * @return boolean True if the hash was set correctly, false otherwise.
+ */
+function set_hash_for_script( string $handle, string $hash ) : bool {
+	global $wp_scripts;
+	return set_hash_for_asset( $wp_scripts, $handle, $hash );
+}
+
+/**
+ * Set the integrity hash for a stylesheet.
+ *
+ * @param string $handle Unique style handle.
+ * @param string $hash Integrity hash (in format "<algo>-<hash>").
+ * @return boolean True if the hash was set correctly, false otherwise.
+ */
+function set_hash_for_style( string $handle, string $hash ) : bool {
+	global $wp_styles;
+	return set_hash_for_asset( $wp_styles, $handle, $hash );
+}
+
+/**
+ * Set the integrity hash for an asset.
+ *
+ * @param WP_Dependencies $dependencies Dependency registry to use.
+ * @param string $handle Unique asset handle.
+ * @param string $hash Integrity hash (in format "<algo>-<hash>").
+ * @return boolean True if the hash was set correctly, false otherwise.
+ */
+function set_hash_for_asset( WP_Dependencies $dependencies, string $handle, string $hash ) : bool {
+	return $dependencies->add_data( $handle, INTEGRITY_DATA_KEY, $hash );
+}
+
+/**
+ * Output the integrity hash for a script.
+ *
+ * This is automatically added to the `script_loader_tag` filter. Use
+ * `set_hash_for_script()` to set the integrity hash for a script.
+ *
+ * @param string $tag Script HTML tag.
+ * @param string $handle Unique script handle.
+ * @return string Script tag with `integrity` attribute set if available.
+ */
+function output_integrity_for_script( string $tag, string $handle ) : string {
+	$hash = get_hash_for_script( $handle );
+	if ( empty( $hash ) ) {
+		return $tag;
+	}
+
+	// Insert the attribute.
+	$tag = str_replace(
+		"type='text/javascript' src='",
+		sprintf(
+			"type='text/javascript' integrity='%s' src='",
+			esc_attr( $hash )
+		),
+		$tag
+	);
+	return $tag;
+}
+
+/**
+ * Output the integrity hash for a stylesheet.
+ *
+ * This is automatically added to the `style_loader_tag` filter. Use
+ * `set_hash_for_style()` to set the integrity hash for a stylesheet.
+ *
+ * @param string $html Stylesheet HTML tag.
+ * @param string $handle Unique style handle.
+ * @return string Stylesheet tag with `integrity` attribute set if available.
+ */
+function output_integrity_for_style( string $html, string $handle ) : string {
+	$hash = get_hash_for_style( $handle );
+	if ( empty( $hash ) ) {
+		return $html;
+	}
+
+	// Insert the attribute.
+	$html = str_replace(
+		" type='text/css'",
```

Comment:
What's the reason for targeting this string over say the end of the tag eg `/>`? Same for the JS one above.

Also is it worth checking whether the `integrity` attribute is already present?

Reply:
With the script tag in particular, the filtered source contains a bunch of tags, so I wanted to target specifically the main tag if possible.

IMO, I don't think it's worth checking for the integrity attribute, as we should just own it as a first-party solution.

Reply:
Was not aware of that, thought they were printed separately :/